src/rgw/rgw_cache.cc

```diff
@@ -344,3 +359,424 @@ ObjectCache::~ObjectCache()
+      exist = true;
+      { /*LRU*/
+
+	/*get ChunkDataInfo*/
+	eviction_lock.Lock();
+	lru_remove(chdo);
+	lru_insert_head(chdo);
+	eviction_lock.Unlock();
+      }
+    } else {	
+      cache_map.erase(oid);
+      lru_remove(chdo);
+      exist = false;
+    }
+  }
+  cache_lock.Unlock();
+  return exist;
+}
+
+size_t DataCache::random_eviction(){
+
+  int n_entries = 0;
+  int random_index = 0;
+  size_t freed_size = 0;
+  ChunkDataInfo *del_entry;
+  string del_oid, location;
+
+  cache_lock.Lock();
+  n_entries = cache_map.size();
+  if (n_entries <= 0){
+    cache_lock.Unlock();
+    return -1;
+  }
+  srand (time(NULL));
+  random_index = rand()%n_entries;
+  map<string, ChunkDataInfo*>::iterator iter = cache_map.begin();
+  std::advance(iter, random_index);
+  del_oid = iter->first;
+  del_entry =  iter->second;
+  ldout(cct, 20) << "INFO::random_eviction index:"<< random_index << ", free size:0x" << std::hex << del_entry->size << dendl;
+  freed_size = del_entry->size;
+  free(del_entry);
+  del_entry = NULL;
+  cache_map.erase(del_oid); // oid
+  cache_lock.Unlock();
+
+  location = cct->_conf->rgw_datacache_persistent_path + del_oid; /*replace tmp with the correct path from config file*/
+  remove(location.c_str());
+  return freed_size;
+}
+
+size_t DataCache::lru_eviction(){
+
+  int n_entries = 0;
+  int random_index = 0;
+  size_t freed_size = 0;
+  ChunkDataInfo *del_entry;
+  string del_oid, location;
+
+  eviction_lock.Lock();
+  del_entry = tail;
+  lru_remove(del_entry);
+  eviction_lock.Unlock();
+
+  cache_lock.Lock();
+  n_entries = cache_map.size();
+  if (n_entries <= 0){
+    cache_lock.Unlock();
+    return -1;
+    }
+  del_oid = del_entry->oid;
+  ldout(cct, 20) << "Engage1: lru_eviction: oid to remove" << del_oid << dendl;
+  map<string, ChunkDataInfo*>::iterator iter = cache_map.find(del_entry->oid);
+  if (iter != cache_map.end()) {
+    cache_map.erase(del_oid); // oid
+  }
+  cache_lock.Unlock();
+  freed_size = del_entry->size;
+  free(del_entry);
+  location = cct->_conf->rgw_datacache_persistent_path + del_oid; /*replace tmp with the correct path from config file*/
+  remove(location.c_str());
+  return freed_size;
+}
+
+
+void DataCache::remote_io(struct librados::L2CacheRequest *l2request ) {
+  ldout(cct, 20) << "Engage1: Add task to remote IO" << dendl;
+  tp->addTask(new HttpL2Request(l2request, cct));
+}
+
+
+std::vector<string> split(const std::string &s, char * delim) {
+  stringstream ss(s);
+  std::string item;
+  std::vector<string> tokens;
+  while (getline(ss, item, (*delim))) {
+    tokens.push_back(item);
+  }
+  return tokens;
+}
+
+void DataCache::push_l2_request(struct librados::L2CacheRequest *l2request ) {
+  tp->addTask(new HttpL2Request(l2request, cct));
+}
+
+static size_t _l2_response_cb(void *ptr, size_t size, size_t nmemb, void* param) {
+  librados::L2CacheRequest *req = static_cast<librados::L2CacheRequest *>(param);
+  req->pbl->append((char *)ptr, size*nmemb);
```

Comment:
this size, nmemb notation (from calloc) seems slightly odd here

Reply:
> this size, nmemb notation (from calloc) seems slightly odd here

This callback function registered in `curl_easy_setopt(curl_handle, CURLOPT_WRITEFUNCTION, _l2_response_cb);` in line 743 (slightly below)
is similar to the Curl documentation example at https://curl.haxx.se/libcurl/c/CURLOPT_WRITEFUNCTION.html
the `size*nmemb` calculation is the same as in the example.
Since this is part of the L2 flow, have not checked yet by running this code, will do so when we reach the stage of L2 verification.